src/api-service/__app__/onefuzzlib/azure/nsg.py

```diff
@@ -0,0 +1,337 @@
+            code=ErrorCode.UNABLE_TO_CREATE,
+            errors=["Unable to update nsg %s due to %s" % (nsg.name, err)],
+        )
+    return None
+
+
+def delete_nsg(name: str) -> bool:
+    # NSG can be only deleted if no other resource is associated with it
+    resource_group = get_base_resource_group()
+
+    logging.info("deleting nsg: %s %s", resource_group, name)
+    network_client = get_network_client()
+
+    try:
+        network_client.network_security_groups.begin_delete(resource_group, name)
+        return True
+    except HttpResponseError as err:
+        err_str = str(err)
+        if (
+            "cannot be deleted because it is in use by the following resources"
+        ) in err_str:
+            return False
+    except ResourceNotFoundError:
+        return True
+
+    return False
+
+
+def set_allowed(name: str, sources: NetworkSecurityGroupConfig) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot update nsg rules. nsg %s not found" % name],
+        )
+
+    logging.info(
+        "setting allowed incoming connection sources for nsg: %s %s",
+        resource_group,
+        name,
+    )
+    all_sources = sources.allowed_ips + sources.allowed_service_tags
+    security_rules = []
+    # NSG security rule priority range defined here:
+    # https://docs.microsoft.com/en-us/azure/virtual-network/network-security-groups-overview
+    min_priority = 100
+    # NSG rules per NSG limits:
+    # https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits?toc=/azure/virtual-network/toc.json#networking-limits
+    max_rule_count = 1000
+    if len(all_sources) > max_rule_count:
+        return Error(
+            code=ErrorCode.INVALID_REQUEST,
+            errors=[
+                "too many rules provided %d. Max allowed: %d"
+                % ((len(all_sources)), max_rule_count),
+            ],
+        )
+
+    priority = min_priority
+    for src in all_sources:
+        security_rules.append(
+            SecurityRule(
+                name="Allow" + str(priority),
+                protocol=SecurityRuleProtocol.TCP,
+                source_port_range="*",
+                destination_port_range="*",
+                source_address_prefix=src,
+                destination_address_prefix="*",
+                access=SecurityRuleAccess.ALLOW,
+                priority=priority,  # between 100 and 4096
+                direction="Inbound",
+            )
+        )
+        # Will not exceed `max_rule_count` or max NSG priority (4096)
+        # due to earlier check of `len(all_sources)`.
+        priority += 1
+
+    nsg.security_rules = security_rules
+    return update_nsg(nsg)
+
+
+def clear_all_rules(name: str) -> Union[None, Error]:
+    return set_allowed(name, NetworkSecurityGroupConfig())
```

Comment:
Can you add something like NetworkSecurityGroupConfig.empty()

So it is explicit that we are setting rules to an empty set

Reply:
yes

Reply:
Okay I just updated. Let me know if this is what you want.